python: Add multisim statistics

Sir-NoChill opened the pull request:
This PR enables the passing of statistics back through the multisim module to the driver code.

# Summary

This pull request implements a simple way to pass statistics from the simulations spawned by multisim back to the parent process and addresses intermittent failure to generate statistics files.

# Motivation

In my workflow, I generally create my own `driver.py` file that is a modified version of `src/python/gem5/utils/multisim/__main__.py` and run my simulations using this file instead of the standard multisim file. I also have a caching mechanism implemented to decide when to rerun the simulations that I did not include in this PR so far (see feedback requests).

This functionality will be useful to expose to users of gem5 to facilitate easier data analysis while using the simulator.

Sir-NoChill committed:
python: Enable passing of stats back to multisim main

Sir-NoChill committed:
python: Allow parsing of simstat dictionary from a stats file

This commit provides a facility to convert the generated stats.txt
file back into a usable python dictionary that conforms to the
stats API.

powerjg commented:
Thanks for this! Can you commit your driver.py in configs/examples/gem5_library/multism (or something like that) so we can see that file as well?

The need for this feature has come up a few times. I'm interested to dig into how you have implemented it!

cc: @BobbyRBruce

Sir-NoChill committed:
python: Support returning id from multisim

This commit returns the multisim simulation id with the statistics
in order to easily identify the statistics used

Sir-NoChill committed:
python: Added custom driver for multisim

This file provides an example for how to work with the statistic
return values in gem5. Note that the caching mechanism is fairly
rudimentary, I would like to (in the future) validate that the
cached results correspond to the ones you want to run by comparing
the generated .ini files for each simulation.

There should also be a cli arg that allows you to force rerun
all of the simulations, but adding that in a normal way seems to
break everything.

Sir-NoChill commented:
I have added my custom driver, the 'caching' mechanism I have is not very good, I think it could be improved by comparing the config.ini files for the cached sims against any new ones we generate, but that will be a process.

erin-le commented on configs/example/gem5_library/multisim/multisim-analyze-statistics-batch.py:
The caching mechanism looks interesting. I think it would be a good idea to improve it by looking at config.ini files. As a note, I ran into some trouble with caching while testing. I tried running multisim with three workloads using `multisim-analyze-statistics-batch.py` and found that .`cached_results` only stores data from 0-1 simulations, and that re-runs exit even if only one set of simulation statistics is cached.

Sir-NoChill replied:
Hm, thought I fixed that one. I will update

erin-le commented on src/python/m5/ext/pystats/simstat.py:
```diff
@@ -48,12 +51,139 @@ def __init__(
         )
+
+    @classmethod
+    def dictFromFile(cls, filename: pathlib.Path) -> Tuple[Dict, pathlib.Path]:
```
It doesn't look like this is called anywhere else in this change. Can you explain how it should be used?

Sir-NoChill replied:
This was my (half-baked) effort at creating a parser for the stats.txt files. I think I will remove that from this PR.

erin-le reviewed (commented):
@Sir-NoChill: Thank you for your contribution! I see that you previously opened and closed PR 2046. Can you confirm if the description of PR 2046, included below, also applies to this PR? 

```
Summary
This pull request implements a simple way to pass statistics from the simulations spawned by multisim back to the parent process and addresses intermittent failure to generate statistics files.

Motivation
In my workflow, I generally create my own driver.py file that is a modified version of src/python/gem5/utils/multisim/__main__.py and run my simulations using this file instead of the standard multisim file. I also have a caching mechanism implemented to decide when to rerun the simulations that I did not include in this PR so far (see feedback requests).

I believe this functionality would be useful to expose to users of gem5 to facilitate easier data analysis while using the simulator.

Feedback Requests
What would the optimal way to expose this functionality to the user be? Ideally, I think the user would supply an additional python script that performs the analysis using the gem5 python statistics API, but I am not sure how that would look/how to implement that.
Should the multisim module be responsible for caching executions and checking them against the existing configurations? (ie. check the .ini files and see if they change between runs, run simulations only if they do)
```
I'm fine with the change to allow multisim to pass statistics back to the parent process for data analysis and to preserve stats in case stats.txt isn't generated. As of PR 1849, the issue where multisim fails to generate stats files should be fixed, but please reopen or create a new issue if it is something you are still running into. Thank you!

BobbyRBruce commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -233,14 +245,17 @@ def handle_exit(signum, frame):
             while remaining_ids and len(active_processes) < max_num_processes:
                 id_to_run = remaining_ids.pop()
                 try:
+                    parent_pipe, child_pipe = Pipe()
```
I'm a little bit confused here. How do we use `Pipe` to get the statistics? Wouldn't `Pipe` just get us the stdout for each gem5 child process? Doesn't this require each child process to dump their stats to the stdout prior to exiting to work?

If this is the case, can't the parent process just each each of the child processes stats files rather than through this mechanism which does much the same? Or read there stdout and stderr files (i believe these are also in each child process's output directory)

Sir-NoChill replied:
My understanding here is that python will end up pickling the stats object and sending it through the pipe from the child process back to the parent process. I did try doing this with the stats files, but instead of building a parser for them I opted to try and pass the dictionary around directly.

I believe that the way this spawns the pipe means that the stdout is unaffected, my thoughts being that it's defining a new file descriptor to pass this specific output through (in this case the stats dictionary).

This is a very similar mechanism alltogether of consuming the stats.txt files, but it avoids the need to write a parser for them.

BobbyRBruce commented on configs/example/gem5_library/multisim/multisim-analyze-statistics-batch.py:
```diff
@@ -0,0 +1,67 @@
+    except OSError:
+        print("No cached results found, running multisim...")
+
+    # print(multisim_args)
```
You can remove commented out code

BobbyRBruce commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -181,8 +187,13 @@ def _run(module_path: Path, id: str) -> None:
-def run(module_path: Path, processes: Optional[int] = None) -> None:
+def run(
+    module_path: Path, processes: Optional[int] = None
+) -> List[Tuple[str, Dict]]:
```
Could you add a line in the docstring below explaining the structure of this return value? What does it represent? 

Couldn't this not just be `Dict[str, Dict]`. A dictionary mapping ID string to a dictionary of statistics (`Dict[<gem5 Process ID>, <stats>]`).

Sir-NoChill replied:
good point, I will update

Sir-NoChill commented:
@erin-le yes, the goal of that previous PR remains for this one, I can update the description if that's necessary.

Good to see more improvements to multisim coming along!

BobbyRBruce commented:
> @erin-le yes, the goal of that previous PR remains for this one, I can update the description if that's necessary.
> 
> Good to see more improvements to multisim coming along!

I'm not sure if you saw my other comments. I only really require an explanation in this PR as to how this all works. I like some of the stuff in here but I just need to understand it a bit better to proceed.

Sir-NoChill committed:
Merge branch 'gem5:develop' into develop

erin-le commented:
@Sir-NoChill Hello, could you respond to @BobbyRBruce's last comment if you have time? It looks like he's just asking for a general explanation. I think it would also be beneficial if you could update the description of this PR with the information from the last PR, or otherwise make it more comprehensive. Thank you!

Sir-NoChill committed:
Merge branch 'gem5:develop' into develop

Sir-NoChill commented:
Sorry about the delay, I will commit my amendments later today. 

@BobbyRBruce I am proposing passing the in-memory SimStat object through a pipe so that we can use the Python 'native' pickling, as opposed to trying to create a new SimStat from the generated files (I tried it that way, but I encountered problems with the serialization, which I have a very janky workaround for in my own tools).

I will remove the SimStat parser as it is unnecessary for this PR, though I might polish my workaround for another PR.

Sir-NoChill committed:
stdlib: refactor multisim batch statistic handling

Sir-NoChill committed:
stdlib: remove simstat parser

Sir-NoChill committed:
stdlib: revert simstat changes

Sir-NoChill committed:
Merge branch 'develop' into develop

BobbyRBruce commented on configs/example/gem5_library/multisim/multisim-analyze-statistics-batch.py:
```diff
@@ -0,0 +1,45 @@
+import json
```
Every file should have a copyright notice.

```
Copyright (c) <date> <copyright holder>
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are
met: redistributions of source code must retain the above copyright
notice, this list of conditions and the following disclaimer;
redistributions in binary form must reproduce the above copyright
notice, this list of conditions and the following disclaimer in the
documentation and/or other materials provided with the distribution;
neither the name of the copyright holders nor the names of its
contributors may be used to endorse or promote products derived from
this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```

Please add them.

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill commented:
FYI: I removed the caching infrastructure for this PR, if I find I need to modify the gem5 source to make caching work, then I will open a new PR to that effect.

BobbyRBruce commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -211,22 +220,23 @@ def run(module_path: Path, processes: Optional[int] = None) -> None:
                     inform(f"Terminating process {process.name}")
                     process.terminate()
-        sys.exit(0)
+        # sys.exit(0) # since we actually want to return our values
```
Acknowledged. You can remove this the and the import

BobbyRBruce commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -181,8 +187,11 @@ def _run(module_path: Path, id: str) -> None:
 
-def run(module_path: Path, processes: Optional[int] = None) -> None:
+def run(module_path: Path, processes: Optional[int] = None) -> Dict[str, Dict]:
     """Run the simulators specified in the module in parallel.
```
It'd be good to update the documentation string here to explain what this function returns

BobbyRBruce reviewed (changes requested):
I like this new addition. I didn't know you could use 'pipe' in this way.

I just have a few minor comments to address then I'll be ok accepting this for merge.

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill committed:
python: Minor fixes and documentation for #2078

- Added copyright notice
- Eliminated commented code
- Documented return type for multisim `run()` function

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill commented:
@BobbyRBruce can I get your review?

Sir-NoChill committed:
util-docker: Install protobuf from source instead of apt

apt repository for protobuf causes errors with absl being out of
date. Installing from source with bazel mitigates this with minimal
space overhead.

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill commented:
@powerjg @BobbyRBruce Just wanted to bump this re-request for review

BobbyRBruce commented on util/dockerfiles/ubuntu-24.04_all-dependencies/Dockerfile:
```diff
@@ -37,9 +37,6 @@ RUN apt -y update && apt -y upgrade && apt -y install \
     zlib1g-dev \
-    libprotobuf-dev \
-    protobuf-compiler \
-    libprotoc-dev \
```
Let's consider this change in a separate PR.


I believe this PR (that is regarding outputting stats for MultiSim) does not require this protobuf change to be reviewed and merged


This entire file should remain unaltered in this PR

Sir-NoChill replied:
Sounds good

BobbyRBruce reviewed (dismissed):
This is looking good to me. I just would like the protobuf change moved out of this PR. Besides this, I have no other comments/suggestions. Happy to merge this when it's addressed

erin-le commented:
@Sir-NoChill: Hello, sorry about the very delayed review. If you're still interested in working on this PR, I just wanted to let you know that Bobby had reviewed it and requested a small change. Thanks for your contribution to gem5!

Sir-NoChill committed:
Merge branch 'develop' into develop

Sir-NoChill committed:
util-docker: revert changes to protobuf

Sir-NoChill commented:
Apologies, I missed that notification. I have made the requested change.

Would you @erin-le like me to squash this into a more digestible commit?

erin-le commented:
Hey @Sir-NoChill, it's up to you. If you're fine with this PR being squashed into a single commit, then I can just squash merge, in which case you wouldn't need to do anything.

If you think it would be better for this PR to appear as multiple commits which each make distinct changes, then please go ahead and clean up the commits on your end.

In either case, the PR description/top comment becomes the commit message, so please double check to see if you're satisfied with it. Thanks!

Sir-NoChill commented:
I have modified the top-level comment and description, you can feel free to squash merge!

erin-le commented on configs/example/gem5_library/multisim/multisim-analyze-statistics-batch.py:
```diff
@@ -0,0 +1,71 @@
+        "##################### END OF SIMULATIONS ############################"
+    )
+
+    for sim, stats in results:
```
I found that I needed to make this change to get it to work:
```suggestion
    for sim, stats in results.items():
```

erin-le commented on src/python/gem5/utils/multisim/multisim.py:
```diff
@@ -255,13 +268,21 @@ def handle_exit(signum, frame):
+                    if not pipe.poll(0):
+                        continue
+                    stats[id] = pipe.recv()
+
```
I ran into an issue where an empty `stats` list would sometimes be returned. Changing this part of the code as follows fixed it for me though.


```suggestion
                curr_processes = active_processes
                active_processes = []
                for process, pipe, id in curr_processes:
                    if pipe.poll(0):
                        try:
                            stats[id] = pipe.recv()
                        except EOFError:
                            pass
                    if process.is_alive():
                        active_processes.append((process, pipe, id))

```

erin-le reviewed (changes requested):
He @Sir-NoChill, sorry to block this again, but I was doing some final testing before merging and ran into some problems. I left suggested fixes for these problems in the following comments; would you be able to take a look and see if you're okay with the changes? If you're fine with the changes, I can just commit them and get this PR in.

Sir-NoChill commented:
@erin-le Thanks for checking that, these look good to me!

erin-le committed:
configs: fix printing step of multisim stats analysis

This commit modifies the printing step of the multisim statistics analysis file to unpack the dictionary with `items()`. This fixes an error.

erin-le committed:
python: fix MultiSim's run function not returning stats

Previously, MultiSim's run function would sometimes be flaky
and return an empty dict for stats. This commit fixes that issue.

Harshil2107 reviewed (approved):
lgtm

Sir-NoChill commented:
should I worry about the clang format issues? I don't think I touched those...

powerjg commented:
> should I worry about the clang format issues? I don't think I touched those...

No... you clearly didn't touch those files.
